File: 3116-kth-smallest-amount-with-single-denomination-combination/3116-kth-smallest-amount-with-single-denomination-combination.py
import math

class Solution:
    def findKthSmallest(self, coins: list[int], k: int) -> int:
        n = len(coins)
        
        def count(x: int) -> int:
            cnt = 0
            for mask in range(1, 1 << n):
                lcm_val = 1
                bits = 0
                for i in range(n):
                    if (mask >> i) & 1:
                        bits += 1
                        lcm_val = math.lcm(lcm_val, coins[i])
                        if lcm_val > x:
                            break
                if lcm_val <= x:
                    if bits % 2 == 1:
                        cnt += x // lcm_val
                    else:
                        cnt -= x // lcm_val
            return cnt

        low = 1
        high = min(coins) * k

        ans = high
        while low <= high:
            mid = (low + high) // 2
            if count(mid) >= k:
                ans = mid
                high = mid - 1
            else:
                low = mid + 1

        return ans


























        

File: 3116-kth-smallest-amount-with-single-denomination-combination/3116-kth-smallest-amount-with-single-denomination-combination.cpp
class Solution {
public:
    long long solve(long long x, vector<int>& coins) {
        int size = coins.size();
        long long sum = 0;
        for (int mask = 1; mask < (1 << size); mask++) {
            int bits = __builtin_popcountll(mask);  
            int64_t leastComMul = 1; 
            for (int j = 0; j < size; j++) {
                if (mask & (1 << j))
                    leastComMul = lcm(leastComMul, coins[j]);
            }
            if (bits & 1)
                sum += x / leastComMul;
            else
                sum -= x / leastComMul;
        }
        return sum;
    }
    long long findKthSmallest(vector<int>& coins, int k) {
        long long l = 0, h = LLONG_MAX, res = 0;
        sort(coins.begin(), coins.end());
        while (l <= h) {
            long long mid = (l + h) / 2;
            if (solve(mid, coins) >= (long long)k) {
                res = mid, h = mid - 1;
            }
            else l = mid + 1;
        }
        return res;
    }
};